### PR TITLE
Implement query cancellation

### PR DESCRIPTION
**Pull request by fritzm**
Details of this commit can be found in u/jbecla/debugKill.
It is a squashed mix of commits from Daniel and Jacek

### CODE REVIEW

File: core/modules/qdisp/ResponseRequester.h [hunk 1]
```diff
@@ -39,6 +39,14 @@ namespace lsst {
 namespace qserv {
 namespace qdisp {
 
+/// ResponseRequester is an interface that handles result bytes. Tasks are
+/// submitted to an Executive instance naming a resource unit (what resource is
+/// required), a request string (task payload), and a requester (handler for
+/// returning bytes). The requester implements logic to process incoming results
+/// and buffers that are sized to the number of bytes expected in the next
+/// segment of results. The requester instance functions also functions as a
```
Comment (fritzm) [inline core/modules/qdisp/ResponseRequester.h]: typo ("functions also functions")

File: core/modules/qdisp/ResponseRequester.h [hunk 2]
```diff
@@ -101,8 +109,10 @@ class ResponseRequester {
         boost::shared_ptr<CancelFunc> f;
         {
             boost::lock_guard<boost::mutex> lock(_cancelMutex);
-            f.swap(_cancelFunc);
-            _cancelled = true;
+            if(!_cancelled) {
+                f = _cancelFunc;
```
Comment (fritzm) [inline core/modules/qdisp/ResponseRequester.h]: The swap() call above nulls out _cancelFunc.  The new code doesn't do this.  Was that an intentional change?

File: core/modules/qdisp/QueryResource.h [hunk 3]
```diff
@@ -58,7 +58,7 @@ class QueryResource : public XrdSsiService::Resource {
                   ExecStatus& status)
         : Resource(rPath.c_str()),
           _session(NULL),
-          _request(NULL),
+          _request(),
```
Comment (fritzm) [inline core/modules/qdisp/QueryResource.h]: Why this change?  _request should be explicitly init'd since it is a raw pointer

File: core/modules/qdisp/ResponseRequester.h (hunk 2 above)
Comment (jbecla) [inline core/modules/qdisp/ResponseRequester.h]: I believe Daniel removed that because it the _cancelFunc is called only in the _callCancel(), and inside _CallCancel() we now have a protection through the "_cancelled" flag, which ensures that we only call that function once.  But I agree it'll be safer to set the _cancelFunc to null, just as yet another level of protection. I'll make the change.

File: core/modules/qdisp/QueryResource.h (hunk 3 above)
Comment (jbecla) [inline core/modules/qdisp/QueryResource.h]: I agree, I missed that, I am not sure what Daniel was thinking, but I'll fix it.

File: core/modules/qdisp/ResponseRequester.h (hunk 2 above)
Comment (jbecla) [inline core/modules/qdisp/ResponseRequester.h]: I tried that, but I backed off, because for some reason it really upsets Qserv badly to the point it segfaults almost every time. Hmm.
